File: car/app/app-samples/navigation/common/src/main/java/androidx/car/app/sample/navigation/common/app/MainActivity.java
/*
 * Copyright (C) 2021 The Android Open Source Project
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package androidx.car.app.sample.navigation.common.app;

import android.Manifest;
import android.content.ComponentName;
import android.content.Context;
import android.content.Intent;
import android.content.ServiceConnection;
import android.os.Bundle;
import android.os.IBinder;
import android.util.Log;
import android.view.View;
import android.widget.Button;
import android.widget.Toast;

import androidx.activity.ComponentActivity;
import androidx.annotation.Nullable;
import androidx.car.app.connection.CarConnection;
import androidx.car.app.sample.navigation.common.R;
import androidx.car.app.sample.navigation.common.nav.NavigationService;

/**
 * The main app activity.
 *
 * <p>See {@link androidx.car.app.sample.navigation.common.car.NavigationCarAppService} for the
 * app's entry point to the cat host.
 */
public class MainActivity extends ComponentActivity {
    static final String TAG = MainActivity.class.getSimpleName();

    // A reference to the navigation service used to get location updates and routing.
    NavigationService mService = null;

    // Tracks the bound state of the navigation service.
    boolean mIsBound = false;

    // Monitors the state of the connection to the navigation service.
    private final ServiceConnection mServiceConnection =
            new ServiceConnection() {

                @Override
                public void onServiceConnected(ComponentName name, IBinder service) {
                    Log.i(TAG, "In onServiceConnected() component:" + name);
                    NavigationService.LocalBinder binder = (NavigationService.LocalBinder) service;
                    mService = binder.getService();
                    mIsBound = true;
                }

                @Override
                public void onServiceDisconnected(ComponentName name) {
                    Log.i(TAG, "In onServiceDisconnected() component:" + name);
                    mService = null;
                    mIsBound = false;
                }
            };

    @Override
    protected void onCreate(@Nullable Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        Log.i(TAG, "In onCreate()");

        setContentView(R.layout.activity_main);

        // Hook up some manual navigation controls.
        Button startNavButton = findViewById(R.id.start_nav);
        startNavButton.setOnClickListener(this::startNavigation);
        Button stopNavButton = findViewById(R.id.stop_nav);
        stopNavButton.setOnClickListener(this::stopNavigation);

        new CarConnection(this).getType().observe(this,
                this::onConnectionStateUpdate);
    }

    @Override
    protected void onStart() {
        super.onStart();
        Log.i(TAG, "In onStart()");
        bindService(
                new Intent(this, NavigationService.class),
                mServiceConnection,
                Context.BIND_AUTO_CREATE);
        requestPermissions(new String[]{Manifest.permission.ACCESS_FINE_LOCATION}, 1);
    }

    @Override
    protected void onStop() {
        Log.i(TAG, "In onStop(). bound" + mIsBound);
        if (mIsBound) {
            // Unbind from the service. This signals to the service that this activity is no longer
            // in the foreground, and the service can respond by promoting itself to a foreground
            // service.
            unbindService(mServiceConnection);
            mIsBound = false;
            mService = null;
        }
        super.onStop();
    }

    private void onConnectionStateUpdate(Integer connectionState) {
        String message = connectionState > CarConnection.CONNECTION_TYPE_NOT_CONNECTED
                ? "Connected to a car head unit"
                : "Not Connected to a car head unit";
        Toast.makeText(this, message, Toast.LENGTH_LONG).show();
    }

    private void startNavigation(View view) {
        if (mService != null) {
            mService.startNavigation();
        }
    }

    private void stopNavigation(View view) {
        if (mService != null) {
            mService.stopNavigation();
        }
    }
}

File: emoji2/emoji2-emojipicker/samples/src/main/java/androidx/emoji2/emojipicker/samples/MainActivity.kt
/*
 * Copyright 2022 The Android Open Source Project
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package androidx.emoji2.emojipicker.samples

import android.app.Activity
import android.content.Context
import android.os.Bundle
import android.widget.Button
import android.widget.EditText
import android.widget.ToggleButton
import androidx.emoji2.emojipicker.EmojiPickerView
import androidx.emoji2.emojipicker.RecentEmojiAsyncProvider
import androidx.emoji2.emojipicker.RecentEmojiProviderAdapter
import com.google.common.util.concurrent.Futures
import com.google.common.util.concurrent.ListenableFuture

class MainActivity : Activity() {
    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.main)

        val view = findViewById<EmojiPickerView>(R.id.emoji_picker)
        view.setOnEmojiPickedListener {
            findViewById<EditText>(R.id.edit_text).append(it.emoji)
        }

        findViewById<ToggleButton>(R.id.toggle_button)
            .setOnCheckedChangeListener { _, isChecked ->
                if (isChecked) {
                    view.emojiGridColumns = 8
                    view.emojiGridRows = 8.3f
                } else {
                    view.emojiGridColumns = 9
                    view.emojiGridRows = 15f
                }
            }
        findViewById<Button>(R.id.button).setOnClickListener {
            view.setRecentEmojiProvider(
                RecentEmojiProviderAdapter(CustomRecentEmojiProvider(applicationContext))
            )
        }
    }
}

/**
 * Define a custom recent emoji provider which shows most frequently used emoji
 */
internal class CustomRecentEmojiProvider(
    context: Context
) : RecentEmojiAsyncProvider {

    companion object {
        private const val PREF_KEY_CUSTOM_EMOJI_FREQ = "pref_key_custom_emoji_freq"
        private const val RECENT_EMOJI_LIST_FILE_NAME =
            "androidx.emoji2.emojipicker.sample.preferences"
        private const val SPLIT_CHAR = ","
        private const val KEY_VALUE_DELIMITER = "="
    }

    private val sharedPreferences =
        context.getSharedPreferences(RECENT_EMOJI_LIST_FILE_NAME, Context.MODE_PRIVATE)

    private val emoji2Frequency: MutableMap<String, Int> by lazy {
        sharedPreferences.getString(PREF_KEY_CUSTOM_EMOJI_FREQ, null)?.split(SPLIT_CHAR)
            ?.associate { entry ->
                entry.split(KEY_VALUE_DELIMITER, limit = 2).takeIf { it.size == 2 }
                    ?.let { it[0] to it[1].toInt() } ?: ("" to 0)
            }?.toMutableMap() ?: mutableMapOf()
    }

    override fun getRecentEmojiListAsync(): ListenableFuture<List<String>> =
        Futures.immediateFuture(emoji2Frequency.toList().sortedByDescending { it.second }
            .map { it.first })

    override fun recordSelection(emoji: String) {
        emoji2Frequency[emoji] = (emoji2Frequency[emoji] ?: 0) + 1
        saveToPreferences()
    }

    private fun saveToPreferences() {
        sharedPreferences
            .edit()
            .putString(PREF_KEY_CUSTOM_EMOJI_FREQ, emoji2Frequency.entries.joinToString(SPLIT_CHAR))
            .commit()
    }
}
